thinkfan.1
.TH THINKFAN "1" "June 2013" "thinkfan 0.9.1" "thinkfan"
.SH NAME
thinkfan \- A simple fan control program
.SH SYNOPSIS
.SY thinkfan
.OP \-hnqzDd
.OP \-b BIAS
.OP \-c CONFIG
.OP \-s SECONDS
.OP \-p [DELAY]
.YS
.SH DESCRIPTION
Thinkfan sets the fan speed according to temperature limits preconfigured in
\fI/etc/thinkfan.conf\fR. It can read temperatures from three possible
sources:
.IP "\fI/proc/acpi/ibm/thermal\fR"
Which is provided by the thinkpad_acpi kernel module,
.IP "\fI/sys/class/hwmon/*/temp*_input\fR"
Which may be provided by any hwmon drivers, and
.IP "\fIS.M.A.R.T.\fR (since 0.9)"
Which reads the temperature directly from the hard disk using libatasmart.
.P
Note that since 0.9 you can use any sensors of these three types at the same
time. To allow that, the configuration \fBkeywords\fR have been \fBchanged\fR. The
\fIsensor\fR keyword has been deprecated in favor of the new keywords
\fItp_thermal\fR, \fIhwmon\fR and \fIatasmart\fR which mark the following path
as a legacy thinkpad_acpi thermal file, sysfs hwmon file, or a hard disk
device file, respectively.
.P
The fan can be \fI/proc/acpi/ibm/fan\fR or some PWM file in \fI
/sys/class/hwmon\fR. Note that the \fIfan\fR config keyword is \fBdeprecated\fR as
well. Instead, you should use \fItp_fan\fR for a legacy thinkpad_acpi fan file
or \fIpwm_fan\fR for a sysfs PWM file.
.P
See the README file and the example configurations for details on these
changes.
.HP
\fBWARNING\fR: This program does only very basic sanity checking on the
configuration. That means that you can set your temperature limits as insane
as you like.
.P
There are two general modes of operation:
.SS COMPLEX MODE
In complex mode, temperature limits are defined for each sensor thinkfan knows
about. Setting suitable limits for each sensor in your system will probably
require a bit of experimentation and good knowledge about your hardware, but
it's the safest way of keeping each component within its specified temperature
range. See http://www.thinkwiki.org/wiki/Thermal_Sensors for details on
which sensor measures what temperature in a Thinkpad. On other systems you'll
have to find out on your own. See the example configs to learn about the
syntax.
.SS SIMPLE MODE
In simple mode, Thinkfan uses only the highest temperature found in the
system. That may be dangerous, e.g. for hard disks.  That's why you should
provide a correction value (i.e. add 10\-15 \°C) for the sensor that has the
temperature of your hard disk (or battery...). See the example config files
for details about that.
.SH CONFIGURATION
Some example configurations are provided with the source package. For detailed
explanations please read the README file. If you installed thinkfan from a
distribution package, you may find them under \fI/usr/share/doc\fR or wherever
your package manager puts documentation.
.SH OPTIONS
.TP
\fB\-h\fR
Show a short help message
.TP
\fB\-s\fR SECONDS
Maximum seconds between temperature updates (default: 5)
.TP
\fB\-b\fR BIAS
Floating point number (-10 to 30) to control rising temperature exaggeration.
If the temperature increases by more than 2 °C during one cycle, this number
is used to calculate a bias, which is added to the current highest temperature
seen in the system:

 current_tmax = current_tmax + delta_t * BIAS / 10

This means that negative numbers can be used to even out short and sudden
temperature spikes like those seen on some on-DIE sensors. Use DANGEROUS mode
to remove the -10 to +30 limit. Note that you can't have a space between -b
and a negative argument, because otherwise getopt will interpret things like
-10 as an option and fail (i.e. write "-b-10" instead of "-b -10").

Default is 15.0
.TP
\fB\-c\fR FILE
Load a different configuration file (default: /etc/thinkfan.conf)
.TP
\fB\-n\fR
Do not become a daemon and log to terminal instead of syslog
.TP
\fB\-q\fR
Be quiet (no status info on terminal)
.TP
\fB\-z\fR
Assume we don't have to worry about resuming from standby when using the sysfs
interface (see README!)
.TP
\fB\-p\fR [SECONDS]
Use the pulsing-fan workaround (for older Thinkpads). Takes an optional
floating-point argument (0-10s) as depulsing duration. Default 0.5s.
.TP
\fB\-d\fR
Do not read temperature from sleeping disks. Instead, 0 °C is used as that
disk's temperature. This is needed if reading the temperature causes your
disk to wake up unnecessarily.
\fBNote\fR: This option is only available if thinkfan was built with -D USE_ATASMART.
.TP
\fB\-D\fR
DANGEROUS mode: Disable all sanity checks. May damage your hardware!!
.SH SIGNALS
SIGINT and SIGTERM simply interrupt operation and should cause thinkfan to
terminate cleanly.
.P
SIGHUP makes thinkfan reload its config. If there's any problem with the new
config, we keep the old one.
.P
SIGUSR1 causes thinkfan to dump all currently known temperatures either to
syslog, or to the console (if running with the -n option).
.SH SEE ALSO
git://git.code.sf.net/p/thinkfan/code

.SH BUGS
If you have any problems with thinkfan, please go to the help forum at sf.net:
.RS
.RE
http://sourceforge.net/projects/thinkfan/forums/forum/905019.
.P
There's a bugtracker at
.RS
.RE
http://sourceforge.net/tracker/?group_id=249873&atid=2416828.



NEWS
thinkfan 0.9.1
==============

This is a minor update which adds a small but useful feature:

On receipt of a SIGUSR1, thinkfan will now print the current known
temperatures. See the manpage for details. This is the only change in
functionality over the previously released 0.9_beta2.


thinkfan 0.9
============

This took a while, and it comes with lots of changes, so brace yourself.

1) S.M.A.R.T via libatasmart
----------------------------
Thinkfan can now read temperatures directly from hard disks via libatasmart.
To enable it, you need to compile with -DUSE_ATASMART.

2) cmake
--------
I've switched to cmake. Don't ask me why, I think it's just more convenient.
Details are in the README.

3) Multiple sensor inputs
-------------------------
When using libatasmart, you obviously want to have other temperature inputs as
well. Thus, thinkfan now allows temperature inputs (i.e. legacy thinkpad_acpi
procfs, sysfs hwmon and atasmart) to be arbitrarily combined. In order to
clean up the config syntax, the "sensor" and "fan" keywords have been
deprecated in favor of other, interface-specific keywords. See README and
example configs for details.

4) Various bugfixes
-------------------
The depulsing hack is now only applied if the fan is actually running (thx
yuszuv). Various sysfs bugs have been fixed and a rare race condition when
resuming from suspend can now be worked around with DANGEROUS mode.



thinkfan 0.8
============

1) Complex Temperature Limits
-----------------------------
Thinkfan now allows you to specify temperature limits for each sensor
individually. Configuring this requires more tweaking and experimentation,
but's definitely worth the effort, since it allows you much more fine-grained
control over your fan's behaviour. See README and the example configs for
details.

2) Arbitrary Strings as Fan Levels
----------------------------------
Now you can specify arbitrary strings as fan levels. This allows you to use
things like "level auto" or "level disengaged" in certain temperature ranges.
The sanity checks will filter out anything that isn't a known sensible string,
but as usual, you can use dangerous mode to disable them.



thinkfan 0.7
============

1) Correction Values
--------------------
Temperatures that are perfectly fine for your CPU may already be dangerous for
most hard disks.

For simplicity's sake, thinkfan uses only the highest temperature found in the
system, but that'll most likely never be your harddisk. Thus you can now
specify a correction value for any sensor. This value is added to the actual
temperature and can be used to give that sensor a greater chance of being
considered in the fan speed decision.
I know this is quite unintuitive, but it's the least intrusive way of handling
this (for now). In a later release, I might extend the config syntax to allow
for sensor-specific temperature limits.

So please do check out the updated README and example configs, and modify your
config to save your harddisk from premature death. :-/

2) Config Syntax
----------------
Thinkfan now has a shiny new (as in cheesy, but better than scanf) parser for
the config files. It was actually already used in 0.6.5. However in this
release it was changed to be much tighter on the config syntax. In all
previous releases, it just skipped anything it didn't reconize. You could thus
have arbitrary garbage in your config file and thinkfan would just ignore it.
These times are over. Thinkfan now rejects anything that isn't a proper
statement or a comment. You can get the old behaviour back by using the -D
(DANGEROUS) option.


thinkfan.c
/*************************************************************************
 * thinkfan version 0.9 -- copyleft 08-2013, Victor Mataré
 *
 * thinkfan is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * thinkfan is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with thinkfan.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 * Although it's generally considered bad style, I do make heavy use of
 * global variables, hoping to minimize memory access in the inner loop.
 * If you know better, please drop me a line. I haven't yet bothered looking
 * at the difference it makes in machine language.
 *************************************************************************/
#include "globaldefs.h"
#include <string.h>
#include <syslog.h>
#include <stdio.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <stdlib.h>
#include <time.h>
#include "thinkfan.h"
#include "message.h"
#include "config.h" // The system interface is part of the config now...

volatile int interrupted;
unsigned int sleeptime, tmp_sleeptime;

#define set_fan cur_lvl = config->limits[lvl_idx].level; \
	if (!quiet && nodaemon) \
	report(LOG_DEBUG, LOG_DEBUG, MSG_DBG_T_STAT); \
	config->setfan();

int run();

/* Return TRUE if *at least one* upper limit has been reached */
int complex_lvl_up() {
	int i;

	for (i=0; likely(i < num_temps); i++)
		if (unlikely(temps[i] >= config->limits[lvl_idx].high[i])) {
#ifdef DEBUG
			triggered_tidx = i;
#endif /* DEBUG */
			return TRUE;
		}
	return FALSE;
}

/* Return TRUE if *all* lower limits have been reached */
int complex_lvl_down() {
	int i;

	for (i=0; likely(i < num_temps && temps[i] <= config->limits[lvl_idx].low[i]); i++);
	if (unlikely(i >= config->limit_len)) return TRUE;
	return FALSE;
}

int simple_lvl_up() {
	if (unlikely(*b_tmax >= config->limits[lvl_idx].high[0])) return TRUE;
	return FALSE;
}

int simple_lvl_down() {
	if (unlikely(*b_tmax <= config->limits[lvl_idx].low[0])) return TRUE;
	return FALSE;
}

void get_temps() {
	tmax = -128;
	tempidx = 0;
	sensoridx = 0;
	found_temps = 0;
	for (sensoridx = 0; likely(sensoridx < config->num_sensors); sensoridx++) {
		config->sensors[sensoridx].get_temp();
	}
	if (unlikely(found_temps < config->used_temps)) {
		report(LOG_ALERT, LOG_ALERT, MSG_ALERT_SENSOR);
		errcnt |= ERR_T_GET;
	}
}



/***********************************************************
 * This is the main routine which periodically checks
 * temperatures and adjusts the fan according to config.
 ***********************************************************/
int fancontrol() {
	float bias=0;
	int diff=0, i, last_errcnt = 0;
	int wt = watchdog_timeout;

	tmp_sleeptime = sleeptime;

	prefix = "\n"; // makes the output more readable
	config->init_fan();
	if (errcnt) return errcnt;

	prefix = "\n"; // It is set to "" by the output macros

	// Set initial fan level...
	lvl_idx = config->num_limits - 1;

	for (i=0; i < num_temps; i++)
		if (temps[i] > tmax) tmax = temps[i];

	/**********************************************
	 * Main loop. This is the actual fan control.
	 **********************************************/
	while(likely(!interrupted && !errcnt)) {

		last_tmax = tmax;
		// depending on the command line, this might also call depulse()
		get_temps();

		// If temperature increased by more than 2 °C since the
		// last cycle, we try to react quickly.
		diff = tmax - last_tmax;
		if (unlikely(diff >= 2)) {
			bias = ((float)diff * bias_level);
			if (tmp_sleeptime > 2) tmp_sleeptime = 2;
		}
		else {
			if (unlikely(diff < 0)) {
				tmp_sleeptime = sleeptime;
				bias = 0;
			}
			else if (unlikely(tmp_sleeptime < sleeptime)) tmp_sleeptime++;
		}
		*b_tmax = tmax + bias;

		// determine appropriate fan level and activate it
		if (unlikely(lvl_idx < config->num_limits - 1 && config->lvl_up())) {
			while (likely(lvl_idx < config->num_limits - 1 && config->lvl_up())) lvl_idx++;
			set_fan;
		}
		else if (unlikely(lvl_idx > 0 && config->lvl_down())) {
			while (likely(lvl_idx > 0 && config->lvl_down())) lvl_idx--;
			set_fan;
			tmp_sleeptime = sleeptime;
		}

		sleep(tmp_sleeptime); // state-dependant sleeptime

		// Write current fan level to IBM_FAN one cycle before the watchdog
		// timeout ends, to let it know we're alive.
		if (unlikely((wt -= tmp_sleeptime) <= sleeptime)) {
			config->setfan();
			wt = watchdog_timeout;
		}

		// slowly reduce the bias
		if (unlikely(bias != 0)) {
			if (likely(bias > 0)) {
				if (bias < 0.5) bias = 0;
				else bias -= bias/2 * bias_level;
			}
			else {
				if (bias > -0.5) bias = 0;
				else bias += bias/2 * bias_level;
			}
		}
		if (unlikely(!chk_sanity)) {
			// In DANGEROUS mode, we exit only after MAXERR consecutive errors.
			if (unlikely(errcnt)) last_errcnt++;
			else last_errcnt = 0;
			if (likely(last_errcnt <= MAXERR)) errcnt = 0;
		}
	}
	return errcnt;
}

void sigHandler(int signum) {
	char *tstat;
	switch(signum) {
	case SIGHUP:
	case SIGINT:
	case SIGTERM:
		interrupted = signum;
		break;
	case SIGUSR1:
		tstat = get_tstat();
		report(LOG_INFO, LOG_INFO, tstat);
		free(tstat);
		if (nodaemon) fputs("\n", stderr);
		break;
	}
}


/***************************************************************
 * Main function:
 * Scan for arguments, set options and initialize signal handler
 ***************************************************************/
int main(int argc, char **argv) {
	int opt, ret;
	char *invalid = "";
	struct sigaction handler;

	rbuf = NULL;
	depulse_tmp = 0;
	sleeptime = 5;
	quiet = FALSE;
	chk_sanity = TRUE;
	bias_level = 1.5f;
	ret = 0;
	config_file = "/etc/thinkfan.conf";
	nodaemon = FALSE;
	errcnt = 0;
	resume_is_safe = FALSE;
	prefix = "\n";
	oldpwm = NULL;
	config = NULL;
	lvl_idx = 0;
	last_tmax = 0;
	tmax = 0;
	temps = NULL;
	cur_lvl = NULL;
	tempidx = 0;
	triggered_tidx = -1;

	char *optstring = "c:s:b:p::hnqDz"
#ifdef USE_ATASMART
			"d";
	dnd_disk = FALSE;
#else
	;
#endif

	openlog("thinkfan", LOG_CONS, LOG_USER);
	syslog(LOG_INFO, "thinkfan " VERSION " starting...");

	interrupted = 0;
	memset(&handler, 0, sizeof(handler));
	handler.sa_handler = sigHandler;
	if (sigaction(SIGHUP, &handler, NULL) \
	 || sigaction(SIGINT, &handler, NULL) \
	 || sigaction(SIGTERM, &handler, NULL) \
	 || sigaction(SIGUSR1, &handler, NULL)) perror("sigaction");

	while ((opt = getopt(argc, argv, optstring)) != -1) {
		switch(opt) {
		case 'h':
			fprintf(stderr, MSG_USAGE);
			return 0;
			break;
#ifdef USE_ATASMART
		case 'd':
			dnd_disk = TRUE;
			break;
#endif
		case 'n':
			nodaemon = 1;
			break;
		case 'c':
			config_file = optarg;
			break;
		case 'q':
			quiet = 1;
			break;
		case 'D':
			chk_sanity = 0;
			break;
		case 'z':
			resume_is_safe = !0;
			break;
		case 's':
			sleeptime = (unsigned int) strtol(optarg, &invalid, 0);
			if (*invalid != 0) {
				report(LOG_ERR, LOG_ERR, MSG_ERR_OPT_S);
				report(LOG_ERR, LOG_INFO, MSG_USAGE);
				ret = 1;
				goto fail;
			}
			break;
		case 'b':
			bias_level = strtof(optarg, &invalid);
			if (*invalid != 0) {
				report(LOG_ERR, LOG_ERR, MSG_ERR_OPT_B);
				report(LOG_ERR, LOG_INFO, MSG_USAGE);
				ret = 1;
				goto fail;
			}
			if (bias_level < -10 || bias_level > 30) {
				report(LOG_ERR, LOG_WARNING, MSG_ERR_OPT_B);
				if (chk_sanity) {
					report(LOG_ERR, LOG_INFO, MSG_USAGE);
					ret = 1;
					goto fail;
				}
			}
			bias_level /= 10;
			break;
		case 'p':
			if (optarg) {
				depulse_tmp = strtof(optarg, &invalid);
				if (*invalid != 0 || depulse_tmp > 10 || depulse_tmp < 0) {
					report(LOG_ERR, LOG_ERR, MSG_ERR_OPT_P);
					report(LOG_ERR, LOG_INFO, MSG_USAGE);
					ret = 1;
					goto fail;
				}
			}
			else depulse_tmp = 0.5f;
			depulse = (useconds_t) (1000*1000 * depulse_tmp);
			break;
		default:
			fprintf(stderr, MSG_USAGE);
			return 1;
		}
	}

	if (sleeptime > 15) {
		report(LOG_ERR, LOG_WARNING, MSG_WRN_SLEEPTIME_15);
		report(LOG_ERR, LOG_INFO, MSG_INF_SANITY);
		if (chk_sanity) {
			ret = 1;
			goto fail;
		}
	}
	else if (sleeptime < 1) {
		report(LOG_ERR, LOG_WARNING, MSG_WRN_SLEEPTIME_1);
		report(LOG_ERR, LOG_INFO, MSG_INF_SANITY);
		if (chk_sanity) {
			ret = 1;
			goto fail;
		}
	}
	watchdog_timeout = sleeptime * 6;

	ret = run();

fail:
	free_config(config);
	free(rbuf);
	free(oldpwm);
	free(temps);
	return ret;
}

/********************************************************************
 * Outer loop. Handles signal conditions, runtime cleanup and config
 * reloading
 ********************************************************************/
int run() {
	int ret = 0, childpid;
	struct tf_config *newconfig=NULL;
	FILE *pidfile;
	rbuf = malloc(sizeof(char) * 128);

	prefix = "\n";

	if (chk_sanity && ((pidfile = fopen(PID_FILE, "r")) != NULL)) {
		fclose(pidfile);
		report(LOG_ERR, LOG_WARNING, MSG_ERR_RUNNING);
		if (chk_sanity) return ERR_PIDFILE;
	}

	if ((config = readconfig(config_file)) == NULL) {
		report(LOG_ERR, LOG_ERR, MSG_ERR_CONF_NOFILE);
		return ERR_CONF_NOFILE;
	}

	config->init_fan();
	get_temps();
	ret |= errcnt;
	if (errcnt) goto cleanexit;

	if (depulse) report(LOG_INFO, LOG_DEBUG, MSG_INF_DEPULSE(sleeptime, depulse_tmp));

	// So we try to detect most errors before forking.

	if (!nodaemon) {
		if ((childpid = fork()) != 0) {
			if (!quiet) fprintf(stderr, "Daemon PID: %d\n", childpid);
			return 0;
		}
		if (childpid < 0) {
			perror("fork()");
			ret |= ERR_FORK;
			goto cleanexit;
		}
	}

	if ((pidfile = fopen(PID_FILE, "w+")) == NULL && chk_sanity) {
		report(LOG_ERR, LOG_WARNING, PID_FILE ": %s", strerror(errno));
		ret |= ERR_PIDFILE;
		goto cleanexit;
	}
	fprintf(pidfile, "%d\n", getpid());
	fclose(pidfile);

	while (1) {
		interrupted = 0;
		if ((ret = fancontrol())) break;
		else if (interrupted == SIGHUP) {
			report(LOG_DEBUG, LOG_DEBUG, MSG_DBG_CONF_RELOAD);
			if ((newconfig = readconfig(config_file)) != NULL) {
				free_config(config);
				config = newconfig;
			}
			else report(LOG_ERR, LOG_ERR, MSG_ERR_CONF_RELOAD);
		}
		else if (SIGINT <= interrupted && interrupted <= SIGTERM) {
			report(LOG_WARNING, LOG_INFO, "\nCaught deadly signal. ");
			break;
		}
	}

cleanexit:
	report(LOG_WARNING, LOG_INFO, MSG_INF_TERM);
	config->uninit_fan();

	unlink(PID_FILE);

	return ret;
}

